File: AOM/packages/aom-protocol-ts/src/envelope.ts
import type { AOMCapability, AOMEdge, AOMNode, EvidenceRef } from "./aom.js";
import type { ArtifactInspection } from "./artifact.js";
import type { GatewayDecision, GatewayRequest, GatewayResponse } from "./gateway.js";
import type {
  JsonValue,
  RawAction,
  RawActionResult,
  RawEvent,
  RawRuntimeSnapshot,
  RawStaticSnapshot,
} from "./raw.js";
import type { TargetDescriptor } from "./target.js";

export type ProtocolMessageKind =
  | "request"
  | "response"
  | "event"
  | "snapshot"
  | "action"
  | "action_result"
  | "target_status";

export type ProtocolPayload =
  | { payloadType: "target_descriptor"; payload: TargetDescriptor }
  | { payloadType: "artifact_inspection"; payload: ArtifactInspection }
  | { payloadType: "raw_event"; payload: RawEvent }
  | { payloadType: "raw_static_snapshot"; payload: RawStaticSnapshot }
  | { payloadType: "raw_runtime_snapshot"; payload: RawRuntimeSnapshot }
  | { payloadType: "raw_action"; payload: RawAction }
  | { payloadType: "raw_action_result"; payload: RawActionResult }
  | { payloadType: "aom_node"; payload: AOMNode }
  | { payloadType: "aom_edge"; payload: AOMEdge }
  | { payloadType: "aom_capability"; payload: AOMCapability }
  | { payloadType: "evidence_ref"; payload: EvidenceRef }
  | { payloadType: "gateway_request"; payload: GatewayRequest }
  | { payloadType: "gateway_decision"; payload: GatewayDecision }
  | { payloadType: "gateway_response"; payload: GatewayResponse }
  | { payloadType: "json"; payload: JsonValue };

export interface ProtocolMessage {
  messageId: string;
  kind: ProtocolMessageKind;
  correlationId?: string;
  payload: ProtocolPayload;
}

export function createProtocolMessage(
  messageId: string,
  kind: ProtocolMessageKind,
  payload: ProtocolPayload,
  correlationId?: string,
): ProtocolMessage {
  return correlationId === undefined
    ? { messageId, kind, payload }
    : { messageId, kind, payload, correlationId };
}

File: AOM/crates/aom-protocol-rs/src/envelope.rs
use crate::{
    AOMCapability, AOMEdge, AOMNode, ArtifactInspection, EvidenceRef, GatewayDecision,
    GatewayRequest, GatewayResponse, RawAction, RawActionResult, RawEvent, RawRuntimeSnapshot,
    RawStaticSnapshot, TargetDescriptor,
};
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum ProtocolMessageKind {
    Request,
    Response,
    Event,
    Snapshot,
    Action,
    ActionResult,
    TargetStatus,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case", tag = "payloadType", content = "payload")]
pub enum ProtocolPayload {
    TargetDescriptor(TargetDescriptor),
    ArtifactInspection(ArtifactInspection),
    RawEvent(RawEvent),
    RawStaticSnapshot(RawStaticSnapshot),
    RawRuntimeSnapshot(RawRuntimeSnapshot),
    RawAction(RawAction),
    RawActionResult(RawActionResult),
    AomNode(AOMNode),
    AomEdge(AOMEdge),
    AomCapability(AOMCapability),
    EvidenceRef(EvidenceRef),
    GatewayRequest(GatewayRequest),
    GatewayDecision(GatewayDecision),
    GatewayResponse(GatewayResponse),
    Json(Value),
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ProtocolMessage {
    pub message_id: String,
    pub kind: ProtocolMessageKind,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correlation_id: Option<String>,
    pub payload: ProtocolPayload,
}
